_posts/2023-01-19-blog-new-svg-engine.md

```diff
@@ -0,0 +1,201 @@
+
+Returning to the test-driven approach, the next best target to fix was text rendering, which was working but not pixel-perfect. The legacy SVG engine takes into account the transformation from the text element up to the topmost renderer when computing the effective "on-screen" font size used to select a font for drawing/measuring, during *layout* time. LBSE needed a way to calculate the CTM for a given SVG renderer, up to a given ancestor renderer (or root), taking into account all possible transformation scenarios, including CSS `transform`, `translate`, `rotate`, SVG `transform` attribute, shifts due to `transform-origin`, perspective transformations, and much more. The *same functionality* is required to implement `getCTM()` / `getScreenCTM()`. 
+
+By the end of **August 2022**, <a href="https://commits.webkit.org/253938@main">`SVGLayerTransformComputation` was added</a> that re-used the existing `mapLocalToContainer()` / `TranformState` API to obtain the CTM. The CTM construction and ancestor chain walk - to accumulate the final transformation matrix - is performed by `mapLocalToContainer()` and no longer needs a special, *incomplete* SVG approach: the existing general approach now works for SVG too.
+
+**September 2022** was mostly devoted to bug fixes related to **pixel-snapping**. Outermost \<svg\> elements <a href="https://commits.webkit.org/254314@main">were not always enforcing stacking contexts</a> and <a href="https://commits.webkit.org/254558@main">failed to align to device pixels</a>. All other elements behaved fine with respect to pixel snapping (*not* applied for SVG elements) unless compositing layers were active. In that case, a `RenderLayerBacking` code path was used that unconditionally applied pixel-snapping - <a href="https://commits.webkit.org/254863@main">avoid that for SVG</a>.
+
+By **October 2022** LBSE could <a href="https://commits.webkit.org/255291@main">properly display</a> SVGs embedded into HTML host documents via \<object\> elements -- the size negotiation logic failed to take into account the LBSE-specific renderers before. CSS `background-image` / `list-image` / HTML \<img\> / etc. <a href="https://commits.webkit.org/255625@main">were fixed as well</a>. Zooming and panning support were <a href="https://commits.webkit.org/255727@main">implemented and improved compared to the legacy engine</a>. Along the way an important bug was fixed, which caused percentage-sized documents (e.g. `width: 100%; height: 100%`) that also specify a `viewBox` to always keep the document size, regardless of the zoom level. Upon zooming, only the stroke width enlarged, but not the boundaries of the document, and thus scrollbars never appeared. This was a *major interop* issue that other browsers already fixed in 2014.
```

Comment:
The way this paragraph ends, it reads as though the major interop issue has not yet been fixed.  I think the paragraph means to say that there was an interop issue since 2014 that LBSE fixes, but was not certain enough to rewrite on that basis.  If that is what it means, a rewrite is needed, which I can do if preferred.  Let me know if so.

Reply:
Please do so - the intention was to highlight that the interop bug is gone.

Reply:
Rewritten.  No further issues, Your Honor.